src/main/java/pl/touk/sputnik/Main.java

```diff
@@ -29,20 +29,19 @@ public static void main(String[] args) {
             System.exit(1);
         }
 
-        ConfigurationHolder.initFromFile(commandLine.getOptionValue(CliOption.CONF.getCommandLineParam()));
-        Configuration configuration = ConfigurationHolder.instance();
+        Configuration configuration = ConfigurationBuilder.initFromFile(commandLine.getOptionValue(CliOption.CONF.getCommandLineParam()));
         configuration.updateWithCliOptions(commandLine);
 
         ConnectorFacade facade = getConnectorFacade(configuration);
-        new Engine(facade).run();
+        new Engine(facade, configuration).run();
     }
 
     private static ConnectorFacade getConnectorFacade(Configuration configuration) {
         ConnectorType connectorType = ConnectorType.getValidConnectorType(configuration.getProperty(GeneralOption.CONNECTOR_TYPE));
         ConnectorFacade facade = null;
         try {
-            facade = ConnectorFacadeFactory.INSTANCE.build(connectorType);
-            facade.validate(ConfigurationHolder.instance());
+            facade = ConnectorFacadeFactory.INSTANCE.build(connectorType, configuration);
+            facade.validate(configuration);
```

Comment:
Do we need to validate, have a validate function? Maybe the builder would be the best place to put validation logic, and we can package-protect facade constructors to ensure that these class instances are created by builders. This is a bit of scope-creep, so it's an optional improvement.